# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: netwerk/protocol/http/nsHttpChannel.cpp
# Commit: 6f4ca967aa44
# Full Hash: 6f4ca967aa446cf24f4224576f51803c982f3d7b
# Author: Kershaw Chang <kershaw@mozilla.com>
# Date: 2024-06-14 20:49:45
# Regressor Bug: 1901990
# File Overlap Count: 2
# Description:
#   Bug 1901990 - Introduce a pref for setting a mock HTTPS RR, r=necko-reviewers,valentin
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D213559
# ==============================================================================

diff -r 8a2bdb44382a -r 6f4ca967aa44 netwerk/protocol/http/nsHttpChannel.cpp
--- a/netwerk/protocol/http/nsHttpChannel.cpp	Fri Jun 14 13:27:36 2024 +0000
+++ b/netwerk/protocol/http/nsHttpChannel.cpp	Fri Jun 14 13:42:37 2024 +0000
@@ -643,6 +643,14 @@
 // When TRR is enabled, we always return true, as resolving HTTPS
 // records don't depend on the network.
 static bool canUseHTTPSRRonNetwork(bool* aTRREnabled = nullptr) {
+  // Respect the pref.
+  if (StaticPrefs::network_dns_force_use_https_rr()) {
+    if (aTRREnabled) {
+      *aTRREnabled = true;
+    }
+    return true;
+  }
+
   if (nsCOMPtr<nsIDNSService> dns = mozilla::components::DNS::Service()) {
     nsIDNSService::ResolverMode mode;
     // If the browser is currently using TRR/DoH, then it can
@@ -6453,7 +6461,9 @@
   // This function currently only supports returning DNS_PREFETCH_ORIGIN.
   // Support for the rest of the DNS_* flags will be added later.
 
-  if (!mProxyInfo) {
+  // When network_dns_force_use_https_rr is true, return DNS_PREFETCH_ORIGIN.
+  // This ensures that we always perform HTTPS RR query.
+  if (!mProxyInfo || StaticPrefs::network_dns_force_use_https_rr()) {
     return DNS_PREFETCH_ORIGIN;
   }
 